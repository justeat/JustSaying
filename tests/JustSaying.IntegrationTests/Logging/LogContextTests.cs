using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using JustSaying.IntegrationTests.Fluent;
using JustSaying.Messaging;
using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Shouldly;
using Xunit.Abstractions;
using Xunit.Sdk;

namespace JustSaying.Logging
{
    public class LogContextTests : IntegrationTestBase
    {
        public LogContextTests(ITestOutputHelper outputHelper) : base(outputHelper)
        { }

        [AwsFact]
        public async Task PublishToTopicLogsShouldHaveContext()
        {
            var services = GivenJustSaying()
                .ConfigureJustSaying(
                    (builder) => builder.WithLoopbackTopic<SimpleMessage>(UniqueName));

            var sp = services.BuildServiceProvider();

            var cts = new CancellationTokenSource();

            var publisher = sp.GetRequiredService<IMessagePublisher>();
            await publisher.StartAsync(cts.Token);

            var message = new SimpleMessage();
            await publisher.PublishAsync(message, cts.Token);

            var output = ((TestOutputHelper) OutputHelper).Output;
            output.ShouldMatchApproved(o => o
                .SubFolder("Approvals")
                .WithScrubber(logMessage => ScrubLogs(logMessage, message.Id.ToString())));

            cts.Cancel();
        }

        [AwsFact]
        public async Task PublishToQueueLogsShouldHaveContext()
        {
            var services = GivenJustSaying()
                .ConfigureJustSaying(
                    (builder) => builder.WithLoopbackQueue<SimpleMessage>(UniqueName));

            var sp = services.BuildServiceProvider();

            var cts = new CancellationTokenSource();

            var publisher = sp.GetRequiredService<IMessagePublisher>();
            await publisher.StartAsync(cts.Token);

            var message = new SimpleMessage();
            await publisher.PublishAsync(message, cts.Token);

            var output = ((TestOutputHelper) OutputHelper).Output;
            output.ShouldMatchApproved(o => o
                .SubFolder("Approvals")
                .WithScrubber(logMessage => ScrubLogs(logMessage, message.Id.ToString())));

            cts.Cancel();
        }

        private string ScrubLogs(string message, string messageId)
        {
            message = message.Replace(messageId, "{MessageId}");
            message = message.Replace(UniqueName, "{TestDiscriminator}");

            message = Regex.Replace(message, @"AwsRequestId: .{8}-.{4}-.{4}-.{4}-.{12}", "AwsRequestId: {AwsRequestId}");
            message = Regex.Replace(message, @"(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})Z", "{DateTime}");
            return message;
        }
    }
}
