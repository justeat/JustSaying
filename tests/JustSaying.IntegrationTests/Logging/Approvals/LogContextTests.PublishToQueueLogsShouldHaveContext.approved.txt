[{DateTime}] info: JustSaying.Fluent.QueuePublicationBuilder[0]
      Adding SQS publisher for message type 'JustSaying.TestingFramework.SimpleMessage'.
[{DateTime}] info: JustSaying.Fluent.QueuePublicationBuilder[0]
      Created SQS publisher for message type 'JustSaying.TestingFramework.SimpleMessage' on queue '{TestDiscriminator}'.
[{DateTime}] dbug: JustSaying[0]
      Checking if queue '{TestDiscriminator}' exists
[{DateTime}] dbug: JustSaying[0]
      Checking if queue '{TestDiscriminator}_error' exists
[{DateTime}] info: JustSaying[0]
      Created queue '{TestDiscriminator}_error' with ARN 'arn:aws:sqs:us-east-1:000000000000:{TestDiscriminator}_error'.
[{DateTime}] info: JustSaying[0]
      Creating error queue {TestDiscriminator}_error
[{DateTime}] info: JustSaying[0]
      Created queue '{TestDiscriminator}' with ARN 'arn:aws:sqs:us-east-1:000000000000:{TestDiscriminator}'.
[{DateTime}] info: JustSaying[0]
      Creating queue {TestDiscriminator} attempt number 0
[{DateTime}] info: JustSaying[0]
      Running 1 startup tasks
[{DateTime}] info: JustSaying[0]
      Starting bus with settings: JustSaying.Messaging.Interrogation.InterrogationResult
[{DateTime}] info: JustSaying.Messaging.Channels.SubscriptionGroups.SubscriptionGroupCollection[0]
      Subscription group collection successfully started
[{DateTime}] info: JustSaying[0]
      Starting bus
[{DateTime}] info: JustSaying.Publish[0]
      => AwsRequestId: {AwsRequestId}
      Published message {MessageId} of type SimpleMessage to Queue 'http://localhost:4566/000000000000/{TestDiscriminator}'.
